plugins/hls-tactics-plugin/src/Rewrite.hs

```diff
@@ -0,0 +1,433 @@
+    -> TacticT jdg ext err s m a
+catch (TacticT t) h = tactic2 $ \jdg ->
+  ProofState $ \s sub ok cut raise eff alt ->
+    runProofState (runStateT t jdg)
+      s sub ok cut
+      (\s' err -> runProofState (tacticToBlah jdg $ h err) s' sub ok cut raise eff alt)
+      eff alt
+
+
+tacticToBlah :: jdg -> TacticT jdg ext err s m a -> ProofState ext err s m (a, jdg)
+tacticToBlah jdg (TacticT t) = runStateT t jdg
+
+
+tactic2 :: (jdg -> ProofState ext err s m (a, jdg)) -> TacticT jdg ext err s m a
+tactic2 f = TacticT $ StateT $ f
+
+
+goal :: TacticT jdg ext err s m jdg
+goal = TacticT get
+
+
+throw :: err -> TacticT jdg ext err s m a
+throw err = TacticT $ lift $ ProofState $ \s _ _ _ raise _ _ -> raise s err
+
+
+subgoals
+    :: Monad m
+    => [jdg -> ProofState ext err s m jdg]
+    -> ProofState ext err s m jdg
+    -> ProofState ext err s m jdg
+subgoals [] p = p
+subgoals (t:ts) (ProofState p) =
+  ProofState $ \s sub ok cut raise eff alt ->
+    p
+      s
+      (\s' jdg k ->
+        runProofState (applyCont (subgoals ts . k) $ t jdg)
+          s' sub ok cut raise eff alt
+      )
+      ok cut raise eff alt
+
+
+(<@>)
+    :: Monad m
+    => TacticT jdg ext err s m a
+    -> [TacticT jdg ext err s m a]
+    -> TacticT jdg ext err s m a
+TacticT t <@> ts =
+  TacticT $ StateT $ \jdg ->
+    subgoals (fmap (\(TacticT t') (_, jdg') -> runStateT t' jdg') ts) $
+      runStateT t jdg
+
+
+proof :: MonadExtract ext m => s -> ProofState ext err s m a -> m [Result s err ext]
+proof s p =
+  runProofState p
+    s
+    (\s' _ x -> proof s' . x =<< hole)
+    (\s' ext -> pure $ pure $ Extract s' ext)
+    (pure $ pure $ NoResult)
+    (\_ err -> pure $ pure $ ErrorResult err)
+    (\ma -> do
+      r <- join ma
+      pure r
+    )
+    (liftA2 (<>))
+
+
+rule :: Functor m => Rule jdg ext err s m ext -> TacticT jdg ext err s m ()
+rule r = TacticT $ StateT $ const $ fmap ((),) $ ruleToProofState r
+
+
+ruleToProofState
+    :: Functor m
+    => Rule jdg ext err s m ext
+    -> ProofState ext err s m jdg
+ruleToProofState r = ProofState $ \s sub ok cut raise eff alt ->
+  case r of
+    Pure ext ->
+      ok s ext
+    SubgoalR jdg k ->
+      sub s jdg $ \ext -> ruleToProofState (k ext)
+    EffectR m ->
+      eff $ fmap (\r' -> runProofState (ruleToProofState r') s sub ok cut raise eff alt) m
+    StatefulR f ->
+      let (s', r') = f s
+       in runProofState (ruleToProofState r') s' sub ok cut raise eff alt
+
+
+runTactic
+    :: MonadExtract ext m
+    => s
+    -> jdg
+    -> TacticT jdg ext err s m a
+    -> m [Result s err ext]
+runTactic s jdg (TacticT t) =
+  proof s (flip evalStateT jdg $ t)
+
+
+proofgoals
+    :: Functor m
+    => ([jdg] -> Maybe err)
+    -> s
+    -> ProofState ext err s m jdg
+    -> ([jdg] -> ProofState ext err s m jdg)
+proofgoals f s (ProofState p) =
+  p s
+    (\s' jdg k goals ->
+      ProofState $ \_ sub _ _ _ _ _ ->
+        sub s' jdg $ \ext -> proofgoals f s' (k ext) $ jdg : goals
```

Comment:
We can push `proofgoals` under the subgoal continuation to avoid needing to generate holes, but that means that `f` doesn't run until it's seen all the holes. Yikes!

Reply:
Alternatively, can just say "fuck it" on the hole effects. I guess it's fine?

Reply:
Do holes NEED effects?